pkg/shell/credentials.js

```diff
@@ -0,0 +1,298 @@
+                body.find(".credential-tab").hide().eq(index).show();
+                body.find(".alert").hide();
+            })
+
+            /* Popover help */
+            .on("click", "[data-toggle='popover']", function() {
+                $(this).popover('toggle');
+            })
+
+            /* Dialog is hidden */
+            .on("hide.bs.modal", function() {
+                if (keys) {
+                    $(keys).off();
+                    keys.close();
+                    keys = null;
+                }
+                hide_add_key();
+            })
+
+            /* Dialog is shown */
+            .on("show.bs.modal", function() {
+                keys = credentials.keys_instance();
+
+                $(keys).on("changed", function() {
+                    var key, id, row, rows = { };
+                    var table = $("#credentials-dialog table.credential-listing");
+
+                    table.find("tbody[data-id]").each(function(i, el) {
+                        row = $(el);
+                        rows[row.attr("data-id")] = row;
+                    });
+
+                    var body = table.find("tbody.ssh-key-body").first();
+                    for (id in keys.items) {
+                        if (!(id in rows)) {
+                            row = rows[id] = body.clone();
+                            row.attr("data-id", id)
+                                .show()
+                                .onoff();
+                            table.append(row);
+                        }
+                    }
+
+                    function text(row, field, string) {
+                        var sel = row.find(field);
+                        string = string || "";
+                        if (sel.text() !== string)
+                            sel.text(string);
+                    }
+
+                    for (id in rows) {
+                        row = rows[id];
+                        key = keys.items[id];
+                        if (key) {
+                            text(row, ".credential-label", key.name || key.comment);
+                            text(row, ".credential-type", key.type);
+                            text(row, ".credential-fingerprint", key.fingerprint);
+                            text(row, ".credential-comment", key.comment);
+                            text(row, ".credential-data", key.data);
+                            row.attr("data-name", key.name)
+                                .attr("data-loaded", key.loaded ? "1" : "0")
+                                .find(".btn-onoff-ct")
+                                    .onoff("value", key.loaded || row.hasClass("unlock"))
+                                    .onoff("disabled", !key.name);
+                        } else if (id !== "adding") {
+                            row.remove();
+                        }
+                    }
+                });
+            });
+    }
+
+    module.exports = {
+        setup: setup
+    };
```

Comment:
Assuming this file is only moved code.

Reply:
Moved and changed. This code only is used in the shell and since i was bringing in new deps I didn't want to bloat the shared part unnecessarily.